text/0000-volatile-copy-and-set.md

```diff
@@ -0,0 +1,232 @@
+to explain the memory model, etc, here:
+http://llvm.org/docs/LangRef.html#volatile-memory-accesses.
+We propose modifying existing documentation, and writing new
+docuemntation, referring to the memory model in the C standard
+instead.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Volatile semantics are not well defined by the C standard, but
+that is out of the scope of this proposal.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+The intrinsics operations already exist and have the semantics
+required by operating system implementors and others.
+
+There are several alternatives, each with their own drawbacks:
+
+1.  Continue using the unstable `core_intrinsics` feature and use the
+    existing unstable intrinsics.  However, this ties the programmer
+    to unstable Rust, which is undesirable in some environments.
+2.  Use the existing copy and set interfaces without volatile qualifiers
+    and hope that the compiler does not elide the relevant calls.  While
+    likely workable in practice for most likely scenarios, this could
+    lead to surprising behavior if the compiler ever incorporates
+    sufficiently advanced analyses that allow it to determine that those
+    elisions are possible from its perspective.  Hope is not a strategy.
+3.  Use the foreign function interface to call separately written code
+    in another language that provides the required semantics.  This
+    is inelegant and complicates the build process.
+4.  Hand-code copy and set loops in terms of the existing `write_volatile`
+    function.  This is inelegant, leads to duplicated code, and opens
+    up the possibility of bugs.  For example, compare:
+
+    ```
+        for (i, elem) in some_slice.iter().enumerate() {
+            unsafe {
+                core::ptr::write_volatile(&mut dest[i], *elem);
+            }
+        }
+    ```
+    to,
+    ```
+        unsafe {
+            core::ptr::copy_volatile(some_slice.as_ptr(), dest.as_mut_ptr(), some_slice.len());
+        }
+    ```
+
+Finally, it is important that this proposal not tie the Rust language
+to the semantics of any given implementation, such as thos defined by
+LLVM.  Futher Rust does not yet have a well-defined memory model we can
+refer to for defining volatile behavior, and C does not define volatile
+`memset`, `memcpy` or `memmove` functions.  However, since the existing
+`core::ptr::write_volatile` and `core::ptr::read_volatile` functions
+are implemented in terms of well-defined semantics, it makes sense to
+use similar semantics here.  We therefore specify that
+`copy_volatile`, `copy_nonoverlapping_volatile` and
+`write_bytes_volatile` adopt semantics similar to those of `read_volatile`
+and `write_volatile`: resulting loads and stores cannot be elided, and
+the relative order to loads and stores cannot be reordered with respect
+to one another, though other operations can be reordered with respect
+to volatile operations.
```

Comment:
I am not entirely sure what this long-winding paragraph is supposed to tell me. :)  The open questions around `volatile` are not new, but can't we specify the new operations entirely in terms of the old? That would avoid even having to talk about `volatile` semantics in this PR.

Reply:
I'm not entirely sure what you mean. Do you mean that we should define the semantics of these operations to match those of `write_volatile` and `read_volatile`? Or that we should mention the LLVM semantics? When I put the proto-RFC on internals, I was asked specifically to avoid tying it to LLVM.

Reply:
> Do you mean that we should define the semantics of these operations to match those of write_volatile and read_volatile?

That. Volatile accesses are not new to Rust, so this RFC shouldn't have to deal with the complexity of specifying them.